### ISSUE TITLE
Investigate size optimization

### ISSUE THREAD

**Issue by markerikson** [rtk-incubator/rtk-query#78]
We should review the current contents of the lib, and see how we can optimize the bundle size.

Currently, RTK Query is **38.8K min, 12.8K min+gz**:

https://bundlephobia.com/result?p=@rtk-incubator/rtk-query@0.1.0

![image](https://user-images.githubusercontent.com/1128784/100823199-c6defe00-3421-11eb-82f5-6def77c11bdc.png)

For comparison, RTK itself is **34.2K min, 10.9K min+gz**:

https://bundlephobia.com/result?p=@reduxjs/toolkit@1.5.0

![image](https://user-images.githubusercontent.com/1128784/100823159-b62e8800-3421-11eb-885a-10990e50e143.png)

and React Query is **25.3K min, 7.2K min+gz**:

https://bundlephobia.com/result?p=react-query@2.26.3

So, RTK + RTK Query is significantly heftier.

Note that about a quarter of RTK Query's bundle size right now is Immer.  When we do merge this into RTK itself, at least _some_ of that ought to go away on net, because RTK already uses most of Immer.

**Comment by markerikson** [comment]
I tried creating a new Redux+TS CRA project, and built it:

```
File sizes after gzip:

  51.04 KB  build\static\js\2.078cd3a7.chunk.js
  1.24 KB   build\static\js\main.bb550caa.chunk.js
  933 B     build\static\css\main.247fdf1d.chunk.css
  786 B     build\static\js\runtime-main.ef28eb2a.js
```

I then added RTK Query, and copy-pasted the `postsApi` from the "React Hooks" example sandbox.  After building:

```
File sizes after gzip:

  61.59 KB (+10.55 KB)  build\static\js\2.2102ca47.chunk.js
  1.54 KB (+303 B)      build\static\js\main.06d54f35.chunk.js
  933 B                 build\static\css\main.247fdf1d.chunk.css
  786 B                 build\static\js\runtime-main.ef28eb2a.js
```

I then ran `source-map-explorer` on the output before and after, and here's the visualization:

First, just the `counterSlice`:

![image](https://user-images.githubusercontent.com/1128784/100831862-9e143400-3434-11eb-9f1d-856be6a1871d.png)

Then, `counterSlice` + `postsApi`:

![image](https://user-images.githubusercontent.com/1128784/100831776-6efdc280-3434-11eb-9c16-335c2676b8c2.png)

Observations:

- Definitely overlap in the chunks of Immer being used
  - but adding the patches support costs us
  - as does the hardcoded ES5 plugin, but we're stuck with that until RTK 2.0
- after that, the biggest pieces are `buildSlice`, `buildThunks`, `buildMiddleware`, and `buildHooks`

Not really sure how we'd go about slimming that down.

But, on the other hand, if you're already using RTK, this isn't _that_ much bigger.

**Comment by phryneas** [comment]
We could create one "with-react" and one "without-react" entry point, that would definitely make it smaller for people not using the generated hooks.

Besides that, maybe investigating a "pure ESM, multiple files with includes" build might be interesting (we could try snowpack?), and of course we can do quite a lot of classic code shaving - but right now I'd rather have a little more size if it allows us to iterate faster.

**Comment by markerikson** [comment]
Yeah, I'm not saying we have to go minify everything right now. But, as good web ecosystem citizens (and maintainers trying to compete with other tools), it's something we oughta make sure we cover before 1.0

**Comment by msutkowski** [comment]
This is addressed in #141
